spot_rl_experiments/spot_rl/envs/gaze_env.py

```diff
@@ -3,103 +3,194 @@
-                )
-            else:
-                cmd_id = spot.set_base_position(
-                    x_pos=goal_x, y_pos=goal_y, yaw=goal_heading, end_time=100
-                )
 
-        x, y, yaw = spot.get_xy_yaw()
-        dist = np.linalg.norm(np.array([x, y]) - np.array([goal_x, goal_y]))
-        heading_diff = abs(wrap_heading(goal_heading - yaw))
-        success = dist < 0.3 and heading_diff < np.deg2rad(5)
+class GazeController:
+    """
+    GazeController is used to gaze at, and pick given objects.
 
-    return traj
+    Args:
+        config (Config): Config object
+        spot (Spot): Spot object
 
+    How to use:
+        1. Create a GazeController object
+        2. Call gaze() method with the target object list
+        3. Call shutdown() to stop the robot
 
-def close_enough(pos1, pos2):
-    dist = np.linalg.norm(np.array([pos1[0] - pos2[0], pos1[1] - pos2[1]]))
-    theta = abs(wrap_heading(pos1[2] - pos2[2]))
-    return dist < 0.1 and theta < np.deg2rad(2)
+    Example:
+        config = construct_config(opts=[])
+        spot = Spot("spot_client_name")
+        gaze_target_list = ["apple", "banana"]
+        gaze_controller = GazeController(config, spot)
+        gaze_results = gaze_controller.gaze(gaze_target_list)
+        gaze_controller.shutdown()
+    """
+
+    def __init__(self, config, spot):
+        self.config = config
+        self.spot = spot
+
+        # Setup
+        self.policy = GazePolicy(config.WEIGHTS.GAZE, device=config.DEVICE)
+        self.policy.reset()
+
+        self.gaze_env = SpotGazeEnv(config, spot)
+        self.gaze_env.power_robot()
+
+    def gaze(self, target_object_list, take_user_input=False):
+        """
+        Gaze at the target object list and pick up the objects if specified in the config
+
+        CAUTION: The robot will drop the object after picking it, please use objects that are not fragile
+
+        Args:
+            target_object_list (list): List of target objects to gaze at
+            take_user_input (bool): Whether to take user input for the success of the gaze
+
+        Returns:
+            gaze_success_list (list): List of dictionaries containing the target object name, time taken and success
+        """
+        gaze_success_list = []
+        print(f"Target object list : {target_object_list}")
+        for target_object in target_object_list:
+            observations = self.gaze_env.reset(target_obj_name=target_object)
+            done = False
+            start_time = time.time()
+            self.gaze_env.say(f"Gaze at target object - {target_object}")
+            while not done:
+                action = self.policy.act(observations)
+                observations, _, done, _ = self.gaze_env.step(arm_action=action)
+            self.gaze_env.say("Gaze finished")
+
+            # Ask user for feedback about the success of the gaze and update the "success" flag accordingly
+            user_feedback = True
```

Comment:
the default should be `False`?

Reply:
The default should be `True`, this is if the user thinks gaze worked as expected.

If `take_user_input` is False, then we will never update the `user_feedback` flag.

The "success" flag on line 159 is an AND operation between `grasp_attempted` and `user_feedback`. To ensure this "success" does not get falsely polluted with `False`, we should keep the default value of `user_feedback` as `True`.